piet-cairo/src/text.rs

```diff
@@ -210,233 +401,219 @@ impl TextLayout for CairoTextLayout {
+             * that case the value tells us the number of "characters" in the grapheme.
+             */
+
+            let text = &self.text;
+            let index = index.try_into().unwrap();
+            let trailing = trailing.try_into().unwrap();
+
+            text[index..]
+                .char_indices()
+                .nth(trailing)
+                .map(|(offset, _)| index + offset)
+                .unwrap_or_else(|| text.len())
         };
 
-        // Trailing whitespace is remove for the line
-        let line = &self.text[lm.range()];
+        let (metric_index, metric) = self
+            .line_metrics
+            .iter()
+            .enumerate()
+            .find(|(_, metric)| metric.start_offset <= index && index < metric.end_offset)
+            .unwrap_or_else(|| {
+                /*
+                 * NOTE: Handle out of bounds end of text index.
+                 * An index such that `index == text.len()` is possible but the
+                 * find predicate will not resolve to any metric in that case so
+                 * always return the last metric if the find fails
+                 */
+                let index = self.line_metrics.len() - 1;
+                (index, &self.line_metrics[index])
+            });
+
+        /*
+         * NOTE: So Pango has some weird behavior with hit tests at and beyond the end of
+         * a soft-broken line. This is an attempt to make it match the behavior we expect.
+         * In a nutshell Pango gives us the index of the last character in the line wheras
+         * we expect and index that equals `line_text.len()` even if there is nothing at
+         * the resulting index. Here we are detecting this index and adjusting it if the hit
+         * point appears to be outside the line on the x-axis in order to prevent a moving
+         * the index in the case of a hit test which should legitimately return the original
+         * value (such as a hit between the last character and the one before it)
+         */
+        let index = {
+            let text = &self.text[metric.start_offset..metric.end_offset];
+
+            let logical_width = self
+                .pango_layout
+                .get_line(metric_index.try_into().unwrap())
+                .unwrap()
+                .get_extents()
+                .1
+                .width as f64
+                / PANGO_SCALE;
+            let outside_x = point.x < 0. || point.x > logical_width;
+
+            let contains_text = !text.is_empty();
+            let last_metric = metric_index + 1 >= self.line_metrics.len();
+            let hard_break = matches!(text.as_bytes(), [.., b'\r', b'\n'] | [.., b'\n']);
+            let wrong_index = index == metric.end_offset.saturating_sub(1);
+
+            if contains_text && !last_metric && !hard_break && wrong_index && outside_x {
+                metric.end_offset
+            } else {
+                index
+            }
+        };
```

Comment:
As mentioned on Zulip I've taken another thorough crack at this issue and I would like a visual once over before I start poking at the tests again

Reply:
I've just realized that this doesn't account for `self.pango_offset.x`, I will correct that